src/engines/arps/ARP/audio.cpp

```diff
@@ -1,185 +1,326 @@
+
+  namespace octave_modes {
+
+    // OctaveMode functions
+    NoteVector standard(ArpeggiatorState& state, const NoteArray& input, PlayModeFunc play_mode)
+    {
+      return play_mode(state, input);
+    }
+
+    NoteVector octaveupunison(ArpeggiatorState& state, const NoteArray& input, PlayModeFunc play_mode)
+    {
+      auto notes = play_mode(state, input);
+      auto res = notes;
+      for (auto note : notes) {
+        res.insert(note + 12);
+      }
+      return res;
+    }
+
+    NoteVector fifthunison(ArpeggiatorState& state, const NoteArray& input, PlayModeFunc play_mode)
+    {
+      auto notes = play_mode(state, input);
+      auto res = notes;
+      for (auto note : notes) {
+        res.insert(note + 7);
+      }
+      return res;
+    }
+
+    NoteVector octaveup(ArpeggiatorState& state, const NoteArray& input, PlayModeFunc play_mode)
+    {
+      auto doubled_input = input;
+      auto max_t = input.back().t;
+      for (auto [i, pair] : util::view::indexed(input)) {
+        doubled_input.push_back({static_cast<std::uint8_t>(pair.note + 12), static_cast<std::int8_t>(max_t + 1 + i)});
+      }
+      return play_mode(state, doubled_input);
+    }
+
+    NoteVector octavedownup(ArpeggiatorState& state, const NoteArray& input, PlayModeFunc play_mode)
+    {
+      auto doubled_input = input;
+      for (auto& pair : doubled_input) {
+        pair.note -= 12;
+      }
+      auto max_t = input.back().t;
+      for (auto [i, pair] : util::view::indexed(input)) {
+        doubled_input.push_back({static_cast<std::uint8_t>(pair.note + 12), static_cast<std::int8_t>(max_t + 1 + i)});
+      }
+      return play_mode(state, doubled_input);
+    }
+  } // namespace octave_modes
 
-  void octaveupdown(std::uint8_t rounds, std::uint8_t note, NoteVector& output)
-  {
-    if (rounds % 2 == 0) output.push_back(note - 12);
-    else output.push_back(note + 12);
-  }
 
-  void octavedownup(std::uint8_t rounds, std::uint8_t note, NoteVector& output)
-  {
-    if (rounds % 2 == 0) output.push_back(note + 12);
-    else output.push_back(note - 12);
-  }
 
   // Action Handlers //
   void Audio::action(itc::prop_change<&Props::playmode>, Playmode pm) noexcept
   {
-    switch(pm){
-    case Playmode::up: playmode_func_ = up; break;
-    case Playmode::down: playmode_func_ = down; break;
-    default: playmode_func_ = up; break;
+    using namespace play_modes;
+    switch (pm) {
+      case Playmode::up: playmode_func_ = up; break;
+      case Playmode::down: playmode_func_ = down; break;
+      case Playmode::updown: playmode_func_ = updown; break;
+      case Playmode::downup: playmode_func_ = downup; break;
+      case Playmode::updowninc: playmode_func_ = updowninc; break;
+      case Playmode::downupinc: playmode_func_ = downupinc; break;
+      case Playmode::manual: playmode_func_ = manual; break;
+      case Playmode::chord: playmode_func_ = chord; break;
+      default: playmode_func_ = play_modes::up; break;
```

Comment:
don't use a default case with enums. The value should never be outside the enum set, and no default case means you get a warning if you havent handled all cases (i.e. we add another playmode later)